File: app/src/main/java/com/pixieium/austtravels/auth/SignUpActivity.java
package com.pixieium.austtravels.auth;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.FirebaseDatabase;
import com.pixieium.austtravels.R;
import com.pixieium.austtravels.databinding.ActivitySignInBinding;
import com.pixieium.austtravels.databinding.ActivitySignupBinding;
import com.pixieium.austtravels.home.HomeActivity;
import com.pixieium.austtravels.models.User;

import org.w3c.dom.Text;

import java.util.ArrayList;

public class SignUpActivity extends AppCompatActivity {
    private ActivitySignupBinding mBinding;
    FirebaseAuth mAuth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = ActivitySignupBinding.inflate(getLayoutInflater());
        View view = mBinding.getRoot();
        setContentView(view);

        mAuth = FirebaseAuth.getInstance();
        initSpinnerSemester();
        initSpinnerDepartment();

        FirebaseDatabase.getInstance().getReference("Demo").child("data").setValue("value");

        mBinding.signup.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                createNewUser();
            }
        });


    }


    private void createNewUser(){

        String email = mBinding.eduMail.getEditText().getText().toString();
        String password = mBinding.password.getEditText().getText().toString();
        String userName = mBinding.name.getEditText().getText().toString();
        String semester = mBinding.semester.getEditText().getText().toString();
        String department = mBinding.semester.getEditText().getText().toString();




        if(TextUtils.isEmpty(email)||TextUtils.isEmpty(password)|| TextUtils.isEmpty(userName)||TextUtils.isEmpty(semester)||TextUtils.isEmpty(department))
        {
            if(TextUtils.isEmpty(email))
            {
                mBinding.eduMail.setError("Email is empty");
            }

            if(TextUtils.isEmpty(password))
            {
                mBinding.password.setError("Email is empty");
            }

            if(TextUtils.isEmpty(userName))
            {
                mBinding.name.setError("Email is empty");
            }
            if(TextUtils.isEmpty(semester))
            {
                mBinding.semester.setError("Semester is empty");
            }
            if(TextUtils.isEmpty(department))
            {
              mBinding.department.setError("Department is empty");
            }
        }

        else {
            mAuth.createUserWithEmailAndPassword(email, password)
                    .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                        @Override
                        public void onComplete(@NonNull Task<AuthResult> task) {
                            if (task.isSuccessful()) {
                                // Sign in success, update UI with the signed-in user's information

                                FirebaseUser user = mAuth.getCurrentUser();
                                if(user!=null)
                                {
                                    UserInfo userInfo = new UserInfo(email,userName,semester,department);
                                    FirebaseDatabase.getInstance().getReference("users")
                                            .child(user.getUid())
                                            .setValue(userInfo).addOnFailureListener(new OnFailureListener() {
                                        @Override
                                        public void onFailure(@NonNull Exception e) {
                                            Log.d("fails",e.toString());
                                        }
                                    });
//                                            .addOnCompleteListener(new OnCompleteListener<Void>() {
//                                        @Override
//                                        public void onComplete(@NonNull Task<Void> task) {
//                                            if (task.isSuccessful())
//                                            {
//                                                Intent intent = new Intent(getApplicationContext(), HomeActivity.class);
//                                                startActivity(intent);
//                                            }
//                                        }
//                                    });

                                }



                            } else {
                                // If sign in fails, display a message to the user.
                                Log.d("signUp",task.getException().toString());

                                Toast.makeText(getApplicationContext(), "Authentication failed.",
                                        Toast.LENGTH_SHORT).show();
                                task.getException().printStackTrace();

                            }
                        }
                    });
        }

    }

    private void initSpinnerSemester()
    {
        ArrayAdapter<String>arrayAdapter;
        ArrayList<String> items = new ArrayList<String>(){{
            add("5.2");
            add("5.1");
            add("4.2");
            add("4.1");
            add("3.2");
            add("3.1");
            add("2.2");
            add("2.1");
            add("1.2");
            add("1.1");
        }};

        arrayAdapter = new ArrayAdapter<String>(this,R.layout.item_spinner,items);

        mBinding.semesterDropdown.setAdapter(arrayAdapter);


    }
    private void initSpinnerDepartment()
    {
        ArrayAdapter<String>arrayAdapter;
        ArrayList<String> items = new ArrayList<String>(){{
            add("CSE");
            add("EEE");
            add("CE");
            add("ME");
            add("IPE");
            add("TE");
            add("BBA");
            add("ARCH");
        }};

        arrayAdapter = new ArrayAdapter<String>(this,R.layout.item_spinner,items);

        mBinding.departmentDropdown.setAdapter(arrayAdapter);


    }
}

File: app/src/main/java/com/pixieium/austtravels/auth/SignUpActivity.kt
package com.pixieium.austtravels.auth

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase
import com.pixieium.austtravels.R
import com.pixieium.austtravels.databinding.ActivitySignupBinding
import com.pixieium.austtravels.home.HomeActivity
import com.pixieium.austtravels.models.UserInfo
import kotlinx.coroutines.launch
import java.util.*
import android.app.Activity
import android.content.Context
import android.view.inputmethod.InputMethodManager


class SignUpActivity : AppCompatActivity() {
    private lateinit var mBinding: ActivitySignupBinding
    private val mDatabase: AuthRepository = AuthRepository()
    private lateinit var mAuth: FirebaseAuth

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = ActivitySignupBinding.inflate(layoutInflater)
        val view: View = mBinding.root
        setContentView(view)
        mAuth = FirebaseAuth.getInstance()

        lifecycleScope.launch {
            val semesters: ArrayList<String> = mDatabase.getSemesterInfo()
            val departments: ArrayList<String> = mDatabase.getDeptInfo()

            initSpinnerSemester(semesters)
            initSpinnerDepartment(departments)
        }

        mBinding.signup.setOnClickListener {
            try {
                hideKeyboard()
            } catch (e: Exception) {
                //e.printStackTrace()
            }
            createNewUser()
        }

    }

    private fun startLoading() {
        mBinding.progressBar.visibility = View.VISIBLE
        mBinding.signup.isEnabled = false
    }

    private fun stopLoading() {
        mBinding.progressBar.visibility = View.GONE
        mBinding.signup.isEnabled = true
    }


    private fun hideKeyboard() {
        // hide the keyboard
        val imm: InputMethodManager =
            getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(mBinding.root.windowToken, 0)
    }

    private fun createNewUser() {
        val email = mBinding.eduMail.editText?.text.toString()
        val password = mBinding.password.editText?.text.toString()
        val userName = mBinding.name.editText?.text.toString()
        val semester = mBinding.semester.editText?.text.toString()
        val department = mBinding.department.editText?.text.toString()
        val universityId = mBinding.universityId.editText?.text.toString()
        val userImage = "https://avatars.dicebear.com/api/bottts/${userName}.svg"
        val userInfo =
            UserInfo(email, password, userName, semester, department, universityId, userImage)

        val errorMessage = userInfo.validateInput()

        if (errorMessage != null) {
            Toast.makeText(this, "Please enter your information correctly", Toast.LENGTH_SHORT)
                .show()
        } else {
            startLoading()
            mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(
                    this
                ) { task ->
                    if (task.isSuccessful) {
                        val user = mAuth.currentUser
                        // uses the dicebears http api to get an image
                        // refer to https://avatars.dicebear.com/docs/http-api
                        if (user != null) {
                            lifecycleScope.launch {
                                if (mDatabase.saveNewUserInfo(userInfo, user.uid, user)) {
                                    sentVerificationEmail()

                                } else {
                                    Toast.makeText(
                                        this@SignUpActivity,
                                        "Something went wrong! Couldn't create your profile",
                                        Toast.LENGTH_SHORT
                                    ).show()
                                }
                                stopLoading()
                            }
                        }
                    }
                }
                .addOnFailureListener(this) {
                    stopLoading()
                    Toast.makeText(
                        applicationContext, it.localizedMessage,
                        Toast.LENGTH_SHORT
                    ).show()
                }
        }


    }

    private fun sentVerificationEmail() {
        val currentUser: FirebaseUser? = mAuth.currentUser
        currentUser?.sendEmailVerification()?.addOnCompleteListener(this) { task ->
            if (task.isSuccessful) {
                Toast.makeText(
                    this@SignUpActivity,
                    "Email verification link sent to your email",
                    Toast.LENGTH_SHORT
                ).show()

                Firebase.auth.signOut()
                val intent =
                    Intent(this@SignUpActivity, SignInActivity::class.java)
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK)
                startActivity(intent)
            }

        }
    }

    private fun initSpinnerSemester(items: ArrayList<String>) {
        val arrayAdapter: ArrayAdapter<String> =
            ArrayAdapter(this, R.layout.item_spinner, items)
        mBinding.semesterDropdown.setAdapter(arrayAdapter)
    }

    private fun initSpinnerDepartment(items: ArrayList<String>) {
        val arrayAdapter: ArrayAdapter<String> =
            ArrayAdapter(this, R.layout.item_spinner, items)
        mBinding.departmentDropdown.setAdapter(arrayAdapter)
    }


}